aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,322 @@
+
+  Tensor output_indptr = at::empty({M + 1}, kLong);
+
+  output_indices.resize_({2, nnz});
+  output_values.resize_(nnz);
+
+  LongTensor output_row_indices = output_indices.select(0, 0);
+  LongTensor output_col_indices = output_indices.select(0, 1);
+
+  _csr_matmult(M, N, mat1_indptr.data_ptr<int64_t>(), mat1_col_indices.data_ptr<int64_t>(), mat1_values.data_ptr<scalar_t>(), 
+  mat2_indptr.data_ptr<int64_t>(), mat2_indices.data_ptr<int64_t>(), mat2_values.data_ptr<scalar_t>(), 
+  output_indptr.data_ptr<int64_t>(), output_col_indices.data_ptr<int64_t>(), output_values.data_ptr<scalar_t>());
+  
+  csr_to_coo(M, output_indptr.data_ptr<int64_t>(), output_row_indices.data_ptr<int64_t>());
+}
+
+template <typename scalar_t>
+Tensor fill_with(const Tensor& input, scalar_t fill_value){
+  auto input_indices = input._indices();
+  auto input_values = input._values();
+
+  if (input.size(0) * input.size(1)  ==  input_values.numel()) {
+    return input;
+  }
+  Tensor output = at::empty_like(input);
+
+  auto output_indices = output._indices();
+  auto output_values = output._values();
+  auto n_rows = input.size(0);
+  auto n_cols = input.size(1);
+  auto matrix_size = n_rows * n_cols;
+  
+  auto new_nnz = input.size(0) * input.size(1) + input_values.numel(); 
+  
+  output_indices.resize_({2, new_nnz});
+  output_values.resize_(new_nnz);
+
+  auto input_indices_accessor = input_indices.accessor<int64_t, 2>();
+  auto input_values_accessor = input_values.accessor<scalar_t, 1>();
+
+  auto output_indices_accessor = output_indices.accessor<int64_t, 2>();
+  auto output_values_accessor = output_values.accessor<scalar_t, 1>();
+
+  at::parallel_for(0, matrix_size, 0, [&](int64_t start, int64_t end) {
+    for (auto index = start; index < end; index++) {
+      output_indices_accessor[0][index] = index / n_cols;
+      output_indices_accessor[1][index] = index % n_cols;
+      output_values_accessor[index] = fill_value;
+    }
+  });
+  at::parallel_for(matrix_size, new_nnz, 0, [&](int64_t start, int64_t end) {
+    for (auto index = start; index < end; index++) {
+      int64_t j = index - matrix_size;
+      output_indices_accessor[0][index] = input_indices_accessor[0][j];
+      output_indices_accessor[1][index] = input_indices_accessor[1][j];
+      output_values_accessor[index] = input_values_accessor[j] - fill_value;
+    }
+  });
+  return output;
+}
+
+template <typename scalar_t, short grad_order>
+void sparse_matmul_kernel_grad(Tensor& output, const Tensor& grad, const Tensor& x) {
+  /* 
+    Computes  the backward output  for matrix C = A@B.
+
+    C = A@B 
+      then 
+    A_grad = C_grad @ B^T
+    B_grad = A^T @ C_grad
+
+    if grad_order == 1:
+      output = x^T @ C_grad 
+    else:
+      output = C_grad @ x^T 
+  */
+  Tensor grad_updated = fill_with(grad, /*fill_value = */ scalar_t(1.0));
+  if (grad_order == 1) {
+    sparse_matmul_kernel<scalar_t>(output, x.transpose(0, 1).coalesce(), grad_updated.coalesce());
```

Comment:
you are calling `.transpose` on x here, but it is an expensive operation requiring copies of indices. Why aren't you specifying CUSPARSE_OPERATION_TRANSPOSE instead in a call to cusparse?

Reply:
I've changed this algorithm, please see the description above.